### PR TITLE
Ghash nist

### LINKED ISSUE #16: tracking: GHASH
Implement GHASH as specified in section 6.4 of the [NIST spec](https://nvlpubs.nist.gov/nistpubs/legacy/sp/nistspecialpublication800-38d.pdf)
- [x] #42 
- [x] #43 
- [x] #25 
- [x] #26 
- [x] #52
- [ ] #53

### LINKED ISSUE #42: test: GHASH
write test vectors for G-Hash

### PR DESCRIPTION
**Pull request by 0xJepsen**
Closes #42 
Closes #16

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -98,4 +98,43 @@ mod tests {
...
+
+        // last bit is 1
+        const H_1: [u8; 16] = hex!("00000000000000000000000000000001");
+        const X_1: [u8; 16] = hex!("00000000000000000000000000000001");
+
+        let mut ghash2 = GHash::new(&H_1.into());
+        ghash2.update(&[X_1.into()]);
+        let result2 = ghash2.finalize();
+
+        // test vector of pain
```
Comment (thor314) [inline src/main.rs:126]: life is pain, great pr

File: circuits/aes-gcm/nistgmul.circom [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+// computes a “product” block, denoted X •Y
+template NistGMulByte() {
+
+    signal input X[16];
+    signal input Y[16];
+    signal output out[16];
+
+    // Let R be the bit string 11100001 || 0120. Given two blocks X and Y
+    // byte 0xE1 is 11100001 in binary
+    // var R[16] = [0xE1, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00];
```
Comment (devloper) [inline circuits/aes-gcm/nistgmul.circom:25]: nit: remove

File: circuits/aes-gcm/nistgmul.circom [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+    //       ⎧ Vi >>1           if LSB1(Vi) = 0;
+    //  Vi+1 ⎨ 
+    //       ⎩ (Vi >>1) ⊕ R     if LSB1(Vi) =1.
+    //  
+    component bit[16];
+    component z_i_update[128];
+    component mulx[128];
+    component bytesToBits = BytesToBits(16);
+    bytesToBits.in <== X;
+    signal bitsX[16*8];
```
Comment (devloper) [inline circuits/aes-gcm/nistgmul.circom:49]: remove duplicate signal, just use bytesToBits.out directly in the loop?

File: circuits/aes-gcm/nistgmul.circom [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+    bitsX <== bytesToBits.out;
+    for (var i = 0; i < 128; i++) {
+        // log("i*8 + j", i*8 + j);
```
Comment (devloper) [inline circuits/aes-gcm/nistgmul.circom:52]: remove

File: circuits/aes-gcm/nistgmul.circom [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+        // z_i_update
+        z_i_update[i] = Z_I_UPDATE(16);
+        z_i_update[i].Z <== Z[i];
+        z_i_update[i].V <== V[i];
+        z_i_update[i].bit_val <== bitsX[i];
+        Z[i + 1] <== z_i_update[i].Z_new;
+
+        // mulx to update V
+        mulx[i] = Mulx(16);
```
Comment (devloper) [inline circuits/aes-gcm/nistgmul.circom:61]: whoa, I can see how this introduces a lot of overhead.   I wonder if it's actually needed to be done per bit like this?

File: circuits/aes-gcm/nistgmul.circom [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+        mulx[i].in <== V[i];
+        V[i + 1] <== mulx[i].out;
+    }
+    // 4. Return Z128. 
+    out <== Z[128];
+}
+
+// mul on 1 byte to debug by hand
+template debug_1_byte() {
```
Comment (devloper) [inline circuits/aes-gcm/nistgmul.circom:70]: remove (?)

File: circuits/aes-gcm/nistgmul.circom [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+        mulx[i].in <== V[i];
+        V[i + 1] <== mulx[i].out;
+        
+        log("V[i]", V[i][0]);
+    }
+    out <== Z[8];
+}
+
+// mul on 2 bytes to debug by hand
+template debug_2_bytes() {
```
Comment (devloper) [inline circuits/aes-gcm/nistgmul.circom:106]: remove (?)

File: circuits/aes-gcm/nistgmul.circom [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+    mux.a <== Z;
+    component xorBlock = XORBLOCK(n_bytes);
+    xorBlock.a <== Z;
+    xorBlock.b <== V;
+    mux.b <== xorBlock.out;
+    Z_new <== mux.out;
+}
+
+// multiplexer for arrays of length n
+template ArrayMux(n) {
```
Comment (devloper) [inline circuits/aes-gcm/nistgmul.circom:162]: move to utils?

File: nistgmul.circom [hunk 9]
```diff
@@ -0,0 +1,12 @@
+template BytesToBits(n) {
```
Comment (devloper) [inline nistgmul.circom:1]: delete file

**Review by devloper — APPROVED** [review]
I did a quick review! I'm supportive of merging this given it's correct and the current baseline is not!

**Pull request merged**